### PR TITLE
Optimize Triton runtime with validated launch caching

### PR DESCRIPTION
**Pull request by whjthu**
`pytest` output:

```
·$ pytest -n 16
========================================================= test session starts =========================================================
platform linux -- Python 3.12.3, pytest-9.0.3, pluggy-1.6.0
rootdir: /home/haojie/ntbackend/.merge-ready/triton-runtime-fastpath
configfile: pyproject.toml
plugins: xdist-3.8.0, anyio-4.13.0, cov-7.1.0
16 workers [375 items]    
............................................................................................................................... [ 33%]
......................................................................................................s........................ [ 67%]
........................................................................................................s................       [100%]
============================================= 373 passed, 2 skipped in 167.53s (0:02:47) ==============================================
```
两个 skip 的是需要多卡环境，在单卡环境上测试被自动跳过了。

## 背景与问题

PR #164 将 NineToothed 切换到新的 SSA lowering 编译链路后，Triton kernel 的公开调用路径增加了较多固定开销。

在当前 `master` 中，每次调用 kernel 都会重复执行：

- 参数与 Launch ABI 绑定
- dtype、device、rank、shape 和 stride 校验
- specialization key 计算
- autotune winner 查找
- Python launcher 参数重组
- Triton kernel launch 对象解析

这些操作不会改善 kernel 本身，却会显著增加端到端延迟。对于 Add、Pow、小型 Matmul 等启动受限算子，运行时固定开销可能高于 kernel 执行时间。

原有调优和 prepared launcher 路径还存在两个正确性风险：

- prepared launcher 的内部返回值可能泄露为公开 API 返回值，而不是返回 Launch ABI 声明的输出 tensor。
- 输入和输出共享存储时，多个候选可能在同一用户 buffer 上依次 benchmark，导致前一个候选修改后一个候选的输入。

## 对比版本

| 名称 | 版本 | 说明 |
|---|---|---|
| **Current** | `triton-runtime-fastpath`，基于 `b29a7e1e674cbeaf9e89be5885d218743a8c4d83`，staged patch SHA256：`92a6553900eba3a533fbdd26c81ac0783a7a25c283c43ebe6c4097a8e1e579c5` | 本 PR 准备合并的版本 |
| **Master** | `b29a7e1e674cbeaf9e89be5885d218743a8c4d83` | PR #164 合入后的干净 master |
| **Old Master** | `c9ebd4950a185beed8d4c1db9ff4a1fd133934ae` | PR #164 合入前的 Triton 实现 |

> Current 尚未 commit，因此暂时使用分支名、base commit 和 staged patch SHA256 唯一标识。创建 commit 后，可将 Current 的版本号替换为实际的 PR HEAD commit。

## 修改内容

本 PR 为 Triton runtime 增加经过验证的快速调用路径。

第一次调用仍执行完整检查，并记录：

- 公开参数与 Launch ABI 的绑定关系
- tensor 的 dtype、device、rank、shape 和 stride
- scalar、constexpr 和 keyword 参数
- specialization identity
- 已选择的 autotune candidate
- 已绑定的底层 Triton launcher

后续调用满足相同运行时契约时，直接复用 prepared launcher，避免重复完成上述工作。参数契约发生变化时会回到完整校验路径，不会跳过安全检查。

具体修改包括：

- 缓存已验证的 runtime call contract。
- 缓存已经完成参数绑定的 prepared launch。
- 为常见 tensor 调用建立低成本 specialization identity。
- 缓存并直接激活 autotune winner。
- 保证公开调用始终返回 Launch ABI 声明的输出 tensor。
- 支持动态 shape、动态 stride、keyword 参数和零尺寸 tensor。
- 检测输入输出存储别名。
- 输入输出存在别名时，不在用户数据上依次 benchmark 多个候选。
- 别名场景选择静态排序最高的合法候选，并只执行一次。
- 修复 xdist worker 缓存父目录未创建的问题。

本 PR 不缓存计算结果，不跳过 kernel 执行，也不根据算子名称替换 kernel。

## 性能指标说明

本文比较的是端到端调用延迟，数值越小越好。

表格中的比值定义为：

```text
Current 延迟 / 对比版本延迟
```

因此：

- 小于 `1.0`：Current 更快。
- 等于 `1.0`：性能基本相同。
- 大于 `1.0`：Current 更慢。

例如：

- `Current / Master = 0.50x` 表示 Current 延迟为 Master 的 50%，即快约 50%。
- `Current / Old Master = 1.20x` 表示 Current 延迟为 Old Master 的 1.2 倍，即慢约 20%。

## 测试方法

测试平台：

- GPU：NVIDIA L20
- Compute Capability：SM89
- Backend：Triton
- 所有性能任务使用同一 GPU 文件锁串行执行

每个测试点：

- 使用 4 个轮换输出 buffer
- 预热 20 次
- 测量 5 轮
- 每轮累计测量时间不少于 30 ms
- 同时记录 wall latency 和 CUDA Event latency
- 测量前执行 PyTorch 数值正确性对比

三个版本使用独立的 `HOME`、NineToothed cache、Triton cache 和 CUDA cache。

Old Master 执行 A1/A2 两次测量，两次结果的整体几何平均漂移为 `1.0005x`。

## 整体性能结果

Current 和 Master 共有 43 个可执行性能点。

| 对比 | 几何平均延迟比 | 性能变化 | Current 获胜点 |
|---|---:|---:|---:|
| **Current / Master** | **0.5088x** | **Current 快约 49.1%** | **43/43** |
| **Current / Old Master** | **0.9004x** | **Current 快约 10.0%** | **28/43** |

Current 与 Master 的生成 Triton 源码 SHA256 在 43 个共同可执行点上完全一致：

```text
43/43 source SHA256 identical
```

因此，Current 相对 Master 的提升来自 runtime 调用路径，而不是生成不同的 kernel。

## 分类性能结果

| 算子类别 | 有效点数 | Current / Master | 相对 Master | Current / Old Master | 相对 Old Master |
|---|---:|---:|---:|---:|---:|
| Add | 6 | `0.4370x` | 快 56.3% | `0.9233x` | 快 7.7% |
| Pow | 6 | `0.4397x` | 快 56.0% | `0.9206x` | 快 7.9% |
| Transpose | 6 | `0.5748x` | 快 42.5% | `2.1354x` | 慢 113.5% |
| Softmax | 0/6 | N/A | 当前不可执行 | N/A | 当前不可执行 |
| LayerNorm | 0/6 | N/A | 当前不可执行 | N/A | 当前不可执行 |
| MaxPool | 5 | `0.3569x` | 快 64.3% | `0.3009x` | 快 69.9% |
| Matmul | 5 | `0.6521x` | 快 34.8% | `0.9573x` | 快 4.3% |
| Addmm | 5 | `0.6494x` | 快 35.1% | `0.9909x` | 快 0.9% |
| Conv | 5 | `0.7245x` | 快 27.5% | `1.0547x` | 慢 5.5% |
| Attention | 5 | `0.3844x` | 快 61.6% | `0.6591x` | 快 34.1% |
| **整体** | **43** | **0.5088x** | **快 49.1%** | **0.9004x** | **快 10.0%** |

## 代表性结果

延迟单位为微秒，数值越小越好。

| 测试点 | Old Master | Master | Current | Current / Master | Current / Old Master |
|---|---:|---:|---:|---:|---:|
| Add 1M | 17.559 us | 58.698 us | 14.948 us | `0.255x` | `0.851x` |
| Pow 1M | 17.578 us | 59.870 us | 14.849 us | `0.248x` | `0.845x` |
| Matmul 512³ | 22.471 us | 64.695 us | 16.291 us | `0.252x` | `0.725x` |
| Addmm 512³ | 27.739 us | 80.224 us | 19.693 us | `0.245x` | `0.710x` |
| Attention B1/H4/S256/D64 | 45.237 us | 90.705 us | 21.229 us | `0.234x` | `0.469x` |
| Transpose 4096² | 66.965 us | 396.525 us | 395.993 us | `0.999x` | `5.913x` |

## 正确性验证

额外验证覆盖：

- 4 个轮换输出 buffer
- 每轮变化的输入数据
- NaN 预填充输出
- 动态 shape
- 动态 stride
- keyword 参数
- 零尺寸 tensor
- 输入输出完全别名
- 部分共享存储
- 交错 view
- 公开返回值身份
- kernel 确实被执行

输入输出存在别名时，Current 不会使用同一用户 buffer 依次 benchmark 多个候选，避免前一个候选修改后一个候选的输入。

## 已知限制

本 PR 只优化 Triton runtime，不修改 SSA lowering、布局规划或生成的 kernel。

Master 与 Current 都有 12 个性能矩阵点无法编译：

- 6 个 Softmax
- 6 个 LayerNorm

失败原因为现有 Triton reduction emitter 生成了 block mask 与 scalar pointer 不匹配的 `tl.load`。

这 12 个点：

- 在 Master 和 Current 中失败集合完全一致。
- 不是本 PR 引入的回归。
- 未计入性能几何平均值。
- 将会在下一个PR修复。

此外：

- Transpose 1024 及更大规模仍明显慢于 Old Master。
- 部分中大型 Conv 仍比 Old Master 慢约 4% 到 77%。
- 当前 SSA Conv 单点冷编译约需 4 到 5 分钟，Old Master 约需 1.5 到 2 分钟。

这些问题属于 layout、contraction lowering 和生成源码复杂度，不能通过 runtime fastpath 解决。

### CODE REVIEW

File: tests/test_aot_auto_tuning.py
Comment (voltjia) [inline tests/test_aot_auto_tuning.py:1]: 按理说这个文件不需要改。

**Review by voltjia — CHANGES_REQUESTED** [review]

**Comment by voltjia** [comment]
如果可以的话，可以给 PR 里加入的这些新的 `dataclass` 都设置成 `kw_only=True`，这样方便后面兼容。

**Review by voltjia — APPROVED** [review]
